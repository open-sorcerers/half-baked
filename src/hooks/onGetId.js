import { map, tap, pipe, last } from 'ramda'
import { fork } from 'fluture'
import { trace } from 'xtrace'

import readFileOr from '../utils/readFileOr'
import selectEntityOr from '../utils/selectEntityOr'

const onGetId = ({ STORAGE, CONSTANTS }) => (req, res, next) => {
  const { NO_MATCH: NOT_FOUND } = CONSTANTS
  const entity = (STORAGE.ACCESS_PATH && last(STORAGE.ACCESS_PATH)) || 'id'
  const finish = (x) => (x[NOT_FOUND] ? res.sendStatus(404) : res.json(x))
  pipe(
    readFileOr(`{}`),
    map(
      pipe(
        JSON.parse,
        selectEntityOr({ [NOT_FOUND]: true }, entity, req.params[entity])
      )
    ),
    fork(next)(finish)
  )(STORAGE.BRAIN)
}

export default onGetId
