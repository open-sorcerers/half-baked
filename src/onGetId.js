const { map, pipe, last } = require('ramda')
const { fork } = require('fluture')
const readFileOr = require('./readFileOr')
const selectBySelector = require('./selectBySelector')

const onGetId = ({ STORAGE }) => (req, res, next) => {
  const sel = last(STORAGE.ACCESS_PATH)
  const finish = (x) => res.json(x)
  pipe(
    readFileOr({}),
    map(JSON.parse),
    selectBySelector(sel, req.params[sel]),
    fork(next)(finish)
  )(STORAGE.BRAIN)
}

module.exports = onGetId
