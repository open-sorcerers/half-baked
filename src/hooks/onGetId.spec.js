import path from 'path'
import onGetId from './onGetId'

const NO_MATCH = 'unable-to-find-this'

test('onGetId - missing', (done) => {
  const raw = {
    STORAGE: {
      // ACCESS_PATH: ['data', 'id'],
      BRAIN: '../../half-bakedata.json'
    },
    CONSTANTS: { NO_MATCH }
  }
  const req = {
    params: {
      id: 'lookup'
    }
  }
  const res = {
    sendStatus: (output) => {
      expect(output).toEqual(404)
      done()
    }
  }
  const next = done
  onGetId(raw)(req, res, next)
})

test('onGetId - /:id cool', (done) => {
  const raw = {
    STORAGE: {
      ACCESS_PATH: ['data', 'id'],
      BRAIN: path.resolve(__dirname, '../../half-bakedata.json')
    },
    CONSTANTS: { NO_MATCH }
  }
  const req = {
    params: {
      id: 'cool'
    }
  }
  const res = {
    sendStatus: done,
    json: (output) => {
      expect(output.data).toMatchSnapshot()
      done()
    }
  }
  const next = done
  onGetId(raw)(req, res, next)
})
