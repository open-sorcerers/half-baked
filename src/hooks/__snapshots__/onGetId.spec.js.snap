// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onGetId - /:id cool 1`] = `
Object {
  "id": "cool",
  "this": "is pretty cool",
}
`;
